crates/noirc_evaluator/src/ssa/acir_gen/internal_var_cache.rs

```diff
@@ -76,10 +98,185 @@ impl InternalVarCache {
             .expect("ICE: `NodeId` type cannot be converted into an `InternalVar`")
     }
 
-    pub(super) fn update(&mut self, id: NodeId, var: InternalVar) {
-        self.inner.insert(id, var);
-    }
     pub(super) fn get(&mut self, id: &NodeId) -> Option<&InternalVar> {
         self.inner.get(id)
     }
+
+    // Transform a field element into a witness
+    // It implements a 'get or create' pattern to ensure only one witness is created per element
+    fn const_to_witness(
+        &mut self,
+        value: FieldElement,
+        evaluator: &mut Evaluator,
+        ctx: &SsaContext,
+    ) -> Witness {
+        if let Some(ids) = ctx.constants.get(&value) {
+            //we have a constant node object for the value
+            if !ids.is_empty() {
+                let id = ids.first().unwrap();
+                let var = self.get_or_compute_internal_var_unwrap(*id, evaluator, ctx);
+                if let Some(w) = var.cached_witness() {
+                    return *w;
+                }
+            }
+            // We generate a witness and assigns it
+            let w = evaluator.create_intermediate_variable(Expression::from(value));
+            for &id in ids {
+                let mut cached_var = self.get_or_compute_internal_var_unwrap(id, evaluator, ctx);
+                assert!(cached_var.cached_witness().is_none());
+                cached_var.set_witness(Some(w));
+                self.update(cached_var);
+            }
+            w
+        } else {
+            //if not, we use the constants map
+            let var = self.constants.entry(value).or_insert(InternalVar::from_constant(value));
+            Self::const_to_witness_helper(var, evaluator)
+        }
+    }
+
+    // Helper function which generates a witness for an InternalVar
+    // Do not call outside const_to_witness()
+    fn const_to_witness_helper(var: &mut InternalVar, evaluator: &mut Evaluator) -> Witness {
+        let w = Self::internal_get_or_compute_witness(var, evaluator);
+        if w.is_none() {
+            let witness = expression_to_witness(var.to_expression(), evaluator);
+            var.set_witness(Some(witness));
+        }
+        var.cached_witness().unwrap()
```

Comment:
Can you change unwrap to expect and state why its okay to unrwap here

Reply:
Done